[Li Daoxin] iP

realidx opened the pull request:
# Yuki
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2425s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is FREE!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

realidx committed:
level-0

realidx committed:
level-1

realidx committed:
level-2

realidx committed:
level-3

realidx committed:
level-4

realidx committed:
A-TextUiTesting

realidx committed:
Level-5

realidx committed:
level-6

realidx committed:
branch-level-7

realidx committed:
branch-level-7

realidx committed:
Merge branch 'branch-Level-7'
merge file io into master branch.

realidx committed:
branch-level-8

realidx committed:
Merge branch 'branch-Level-8'
adding Date and Time function to Yuki.

realidx committed:
A-MoreOOP

realidx committed:
A-MoreOOP

realidx committed:
Merge tag 'A-MoreOOP'
Implement A-MoreOOP

realidx committed:
A-Packages

realidx committed:
Merge branch 'A-Packages'
Implement A-Packages

realidx committed:
Merge branch 'add-gradle-support'
Add gradle support.

realidx committed:
Modified buil.gradle

realidx committed:
J-JUnit

realidx committed:
Merge branch 'A-JUnit'

realidx committed:
A-JavaDoc

realidx committed:
Level-9

realidx committed:
Merge branch 'A-CodingStandard'

realidx committed:
Merge branch 'branch-Level-9'

pang-yy commented on src/main/java/yuki/task/Task.java:
I like how clean this bit of code is 👍

pang-yy commented on src/main/java/yuki/Ui.java:
```diff
@@ -0,0 +1,47 @@
+        System.out.println(BORDER);
+    }
+
+    public void showLine() {
+        System.out.println(BORDER + "\n");
+    }
+
+    public void showGoodbye() {
+        System.out.println("Bye. Hope to see you again soon!");
+    }
+
+    public void showError(String message) {
+        System.out.println(message);
+    }
+
+    public String readCommand() {
+        return new java.util.Scanner(System.in).nextLine();
+    }
+
+    public void showLoadingError() {
+        System.out.println("Error loading file. Starting with an empty task list.");
+    }
+
+    /**
+     * Prints the message to the console.
+     *
+     * @param message The message to be printed.
+     */
+    public void print(String message) {
+        System.out.println(message);
+    }
+}
```
Isn't it better for `showLine()`, `showGoodbye()`, etc., to use `print(...)` since it already exists? 🤔

pang-yy commented on src/main/java/yuki/Storage.java:
```diff
@@ -0,0 +1,76 @@
+     *
+     * @param tasks The list of tasks to be saved.
+     */
+    public void save(TaskList<Task> tasks) {
+        try {
+            FileWriter writer = new FileWriter(filePath);
+            for (int i = 0; i < tasks.size(); i++) {
+                writer.write(tasks.get(i).toFileString() + "\n");
+            }
+            writer.close();
+        } catch (Exception e) {
+            System.out.println("An error occurred while saving tasks to file.");
+        }
+    }
```
I like how you separate `save()` and tasks formatting.

pang-yy commented on src/test/java/yuki/task/TodoTest.java:
```diff
@@ -0,0 +1,17 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TodoTest {
+    @Test
+    public void dummyTest(){
+        assertEquals(2, 2);
+    }
+
+    @Test
+    public void anotherDummyTest(){
+        assertEquals(4, 4);
+    }
+}
```
The test cases seem unnecessary. 🤔

pang-yy reviewed (commented):
Overall, I found your code easy to read.

Looks good. 👍 👍 👍

Victoria281 commented on src/main/java/yuki/command/BasicCommand.java:
```diff
@@ -0,0 +1,100 @@
+    public void execute(TaskList<Task> tasks, Ui ui, Storage storage) throws YukiException {
+        switch (this.getCommand(0)) {
+            case "list" -> {
+                ui.showLine();
```
I like that you used switch case here. 
The course suggests a specific style for switch-case formatting, should we align with that?

More information is in the video under W3.5a Implementation → Code Quality → Introduction → What

Victoria281 commented on src/main/java/yuki/command/BasicCommand.java:
```diff
@@ -0,0 +1,100 @@
+                    System.out.println("Task number does not exist");
+                } catch (NumberFormatException e) {
+                    System.out.println("Please enter a valid task number");
+                }
```
The different cases under the switch statement seem to be doing multiple things. Would it improve readability if we abstracted each case into its method for better separation of concerns?

Victoria281 commented on src/main/java/yuki/Yuki.java:
```diff
@@ -0,0 +1,55 @@
+                isExit = c.isExit();
+            } catch (YukiException e) {
+                ui.showError(e.getMessage());
+            }
```
I like how you've handled exceptions here, but maybe it would be better to standardize a custom error message across all cases under YukiException etc. for better user experience?

You could use System.err.println to show the error message in red in the terminal

Victoria281 commented on src/main/java/yuki/YukiException.java:
```diff
@@ -0,0 +1,11 @@
+ * YukiException class deals with exceptions thrown by Yuki.
+ */
+public class YukiException extends Exception{
+    public YukiException(String message) {
+        super(message);
+    }
```
```suggestion
    public YukiException() {
        super("Oh no! Please contact the developer for help!");
    }
```

Victoria281 commented on src/main/java/yuki/command/BasicCommand.java:
```diff
@@ -0,0 +1,100 @@
+                    }
+                    ui.showLine();
+                } catch (IndexOutOfBoundsException e) {
+                    System.out.println("Please enter a keyword to search for");
```
Would it be clearer if we threw a custom exception (e.g., NoKeywordMatchException) here to inform the user that their search led to no results? 

This would improve code clarity by explicitly providing a message to the user

Victoria281 commented on src/main/java/yuki/command/Command.java:
```diff
@@ -0,0 +1,32 @@
+        this.description = description;
+    }
+
+    public String getCommand(int i){
```
Instead of using raw string values for commands, should we use an Enum?  
This would allow us to store all command names in a centralized Enum file, making it easier to manage and update command names.  Additionally, we could add a function in the Enum to normalize user input for handling uppercase and lowercase, ensuring consistent processing  of commands

Victoria281 commented on src/main/java/yuki/TaskList.java:
```diff
@@ -0,0 +1,52 @@
+        taskList = new ArrayList<>();
+    }
+    public TaskList(TaskList<T> storageList) {
+        taskList = storageList.taskList;
```
I like that you're aiming to create a generic type for TaskList. However, I noticed that this constructor directly references taskList from storageList, meaning both instances share the same reference. 

Doesn't this go against the idea of having a generic task list?

Victoria281 commented on src/main/java/yuki/TaskList.java:
```diff
@@ -0,0 +1,52 @@
+     * @param taskNumber Index of the task
+     * @return Task at the specified index
+     */
+    public Task get(int taskNumber) {
+        return (Task) taskList.get(taskNumber);
+    }
```
Since TaskList<T> is generic, should this method return T instead of casting to (Task)? The current approach seems to assume T is always Task.

Victoria281 reviewed (commented):
Great work! Just a few cosmetic changes and coding questions to address.

realidx committed:
adding checkstyle

realidx committed:
Merge branch 'A-Checkstyle'
merge A-Checkstyle

realidx committed:
Level-10

realidx committed:
Merge branch 'Level-10'
merge branch-level-10 to add JavaFX feature

realidx committed:
A-CodeQuality

realidx committed:
Merge tag 'A-CodeQuality'
New increment about code quality.

realidx committed:
A-CI

realidx committed:
Merge pull request #2 from Prog-Neuro-Com/A-CI

add A-CI feature

realidx committed:
refine

realidx committed:
Merge pull request #3 from Prog-Neuro-Com/A-CodeQuality

refine

realidx committed:
Update command handling

realidx committed:
Add undo functionality for the most recent command

- Implemented the ability to undo the most recent command,
  enhancing user experience and error recovery.
- Maintained a history stack to track executed commands
  for easy reversal.
- Ensured the undo operation is safe and does not affect
  other tasks.
- Added test cases to verify the correctness and reliability
  of the undo feature.

This change improves usability by allowing users to revert unintended actions quickly.

realidx committed:
Update deletion function

realidx committed:
Merge branch 'BCD-Extension'

realidx committed:
Update continuous integration

realidx committed:
Add user guide

realidx committed:
Merge branch 'A-UserGuide'

realidx committed:
Add more error handling

realidx committed:
Merge branch 'A-MoreErrorHandling'

realidx committed:
Add more error handling

realidx committed:
Add more error handling

realidx committed:
Update README.md

realidx committed:
Update file storage

realidx committed:
Update file storage

realidx committed:
Update file storage
